src/community/ardrive_contract_oracle.ts

```diff
@@ -0,0 +1,143 @@
+import Arweave from 'arweave';
+import { ContractOracle, ContractReader } from './contract_oracle';
+import { VertoContractReader } from './verto_contract_oracle';
+import { SmartweaveContractReader } from './smartweave_contract_oracle';
+import { RedstoneContractReader } from './redstone_contract_oracle';
+
+// ArDrive Profit Sharing Community Smart Contract
+export const communityTxId = '-8A6RexFkpfWwuyVO98wzSFZh0d6VJuI-buTJvlwOJQ';
+
+const maxReadContractAttempts = 3;
+
+export interface CommunityContractData {
+	votes: [Record<string, unknown>];
+	settings: [string, unknown][];
+	balances: { [tokenHolderAddress: string]: number };
+	vault: { [tokenHolderAddress: string]: { balance: number; start: number; end: number }[] };
+}
+
+/**
+ * Oracle class responsible for retrieving the correct data fields from
+ * the ArDrive Community Contract. This class can utilize several different
+ * contract readers and will fallback to other readers if one fails
+ *
+ * @remarks Will begin fetching data from default contract reader on construction
+ */
+export class ArDriveContractOracle implements ContractOracle {
+	constructor(private readonly arweave: Arweave) {
+		// Get contract data upon construction
+		this.getCommunityContract();
+	}
+
+	/**
+	 * Array of contract readers to use as fall back if one fails
+	 * Uses contract reader at index 0 first then descends down the list
+	 */
+	private contractReaders: ContractReader[] = [
+		new VertoContractReader(),
+		new SmartweaveContractReader(this.arweave),
+		new RedstoneContractReader(this.arweave)
+	];
+	private currentContractReader = 0;
+	private readContractAttempts = 0;
+
+	private communityContract?: CommunityContractData;
+	private contractPromise?: Promise<CommunityContractData>;
+
+	/** Sets the current contract reader to the next reader in descending order  */
+	fallbackToNextContractReader(): void {
+		const nextIndex = this.currentContractReader + 1;
+		const maxIndex = this.contractReaders.length - 1;
+		const firstIndex = 0;
+
+		this.currentContractReader = nextIndex < maxIndex ? nextIndex : firstIndex;
+
+		// eslint-disable-next-line no-console
+		console.log('Falling back to next contract reader..');
+	}
+
+	/**
+	 * Reads the ArDrive Community Contract with the current contract reader
+	 *
+	 * @remarks Will fallback to other contract readers when one fails
+	 */
+	async readContract(txId: string): Promise<unknown> {
+		let contract: unknown;
+
+		try {
+			contract = await this.contractReaders[this.currentContractReader].readContract(txId);
```

Comment:
Comment here to explain that its the contract reader's readContract implementation